### PR TITLE
Particle Swarm Optimisation

### PR DESCRIPTION
**Pull request by VolodyaCO**
## Description

PSO is a SOTA algorithm for many problems, and one of the most common used global optimisation routines. This optimiser is pretty much used for benchmarking purposes for the quantum AI team, as well as for actual applications for the professional services team.

This PR adds the continuous version of PSO, and fixes some performance issues + bugs in the implementation found in [here](https://github.com/zapatacomputing/z-global-optimization/blob/main/src/zquantum/global_optimization/continuous_pso.py). For the time being, I am only interested in this version of PSO, so I will not port the other versions found in https://github.com/zapatacomputing/z-global-optimization.

I need help with the tests because I don't understand the general structure of the tests.

## Please verify that you have completed the following steps

- [x] I have self-reviewed my code.
- [x] I have included test cases validating introduced feature/fix.
- [x] I have updated documentation.

### CODE REVIEW

File: src/orquestra/opt/optimizers/pso/continuous_pso_optimizer.py [hunk 1]
```diff
@@ -0,0 +1,333 @@
...
+        A Bounds object, which can be a scipy.optimize.Bounds object, a sequence of
+        tuples, one tuple being the bonds per parameter, or a tuple of two floats,
+        which are the lower and upper bounds for all parameters.
+
+    Returns
+    -------
+    Union[np.ndarray, Tuple[np.ndarray, np.ndarray]]
+        Lower and upper bounds.
+    """
+    if isinstance(bounds, scipy.optimize.Bounds):
```
[the remark this answers is not in the document]
Comment (mstechly) [reply src/orquestra/opt/optimizers/pso/continuous_pso_optimizer.py:40]: Also, it seems that bounds are used only in init, so perhaps we don't need this function at all.
What about having a function `extract_scale_and_shift(bounds: Bounds)` ?

File: src/orquestra/opt/optimizers/pso/continuous_pso_optimizer.py [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+        self.max_iterations = max_iterations
+        self.max_fevals = max_fevals
+
+        # Other attributes:
+        self.swarm_size = swarm_size
+        self.random_number_generator = np.random.default_rng(seed)
+        self.bounds = _get_bounds_like_array(bounds)
+        self.scale = self.bounds[1] - self.bounds[0]
+        self.shift = self.bounds[0]
+        self.function_at_best_positions = np.ones(swarm_size, dtype=float) * np.infty
```
Comment (mstechly) [inline src/orquestra/opt/optimizers/pso/continuous_pso_optimizer.py:154]: [request] Can we change this to `np.inf`?
See: https://numpy.org/devdocs/reference/constants.html#numpy.infty

File: src/orquestra/opt/optimizers/pso/continuous_pso_optimizer.py [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+    ):
+        """
+        Constructor of the Particle Swarm Optimizer for continuous variables.
+        When you call `.minimize`, the initial positions of the particles can be
+        set using the method `get_initial_params`, which by default samples
+        uniformly from the parameter space.
+
+        Parameters
+        ----------
+        swarm_size : int
```
Comment (mstechly) [inline src/orquestra/opt/optimizers/pso/continuous_pso_optimizer.py:72]: [comment] No need to include type here, as the type is already in the signature. When you have them in two places, you always need to remember to update them in both, and as we know, humans make mistakes.

File: src/orquestra/opt/optimizers/pso/continuous_pso_optimizer.py [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+        Returns
+        -------
+        np.ndarray
+            (N, dimensions) array, where N is the swarm size.
+        """
+        velocities = self.random_number_generator.uniform(
+            0, 1, (self.swarm_size, dimensions)
+        )
+        if self.velocity_bounds is None:
+            self.velocity_bounds = None
```
Comment (mstechly) [inline src/orquestra/opt/optimizers/pso/continuous_pso_optimizer.py:183]: [request] This line is unnecessary, please delete.

File: src/orquestra/opt/optimizers/pso/continuous_pso_optimizer.py [hunk 5]
```diff
@@ -0,0 +1,333 @@
...
+            velocities = 0.5 * (self.scale * velocities - self.shift)
+        else:
+            self.velocity_bounds = _get_bounds_like_array(self.velocity_bounds)
```
Comment (mstechly) [inline src/orquestra/opt/optimizers/pso/continuous_pso_optimizer.py:186]: [request] Could use similar refactor as I mentioned for "regular bounds"

File: src/orquestra/opt/optimizers/pso/continuous_pso_optimizer.py [hunk 6]
```diff
@@ -0,0 +1,333 @@
...
+        self.inertia = inertia
+        self.topology_constructor = topology_constructor
+        self.affinity_towards_best_particle_position = (
+            affinity_towards_best_particle_position
+        )
+        self.affinity_towards_best_swarm_position = affinity_towards_best_swarm_position
+        self.learning_rate = learning_rate
+        self.velocity_bounds = velocity_bounds
+
+    def get_initial_velocities(self, dimensions: int) -> np.ndarray:
```
Comment (mstechly) [inline src/orquestra/opt/optimizers/pso/continuous_pso_optimizer.py:164]: [question] Is this intended to ever be used outside of the class methods, directly by a user?
If no, please change these methods to private by adding `_` before the names.


This applies to all other methods of this class

File: src/orquestra/opt/optimizers/pso/continuous_pso_optimizer.py [hunk 7]
```diff
@@ -0,0 +1,333 @@
...
+        else:
+            self.positions = self.get_initial_params(dimensions)
+            # Set first particle in the swarm to the initial parameters:
+            self.positions[0, :] = initial_params
+        self.best_positions = self.positions.copy()
+        self.topology = self.topology_constructor(dimensions)
+        self.velocities = self.get_initial_velocities(dimensions)
+
+        n_iterations_since_last_improvement = 0
+        best_swarm_value_checkpoint = np.infty
```
Comment (mstechly) [inline src/orquestra/opt/optimizers/pso/continuous_pso_optimizer.py:291]: Please change to `np.inf`, see previous comments.

**Review by mstechly — CHANGES_REQUESTED** [review]
Adding seed control would be also a nice feature, but I think it should be done in a separate PR.

I pointed out some style issues, but overall I'm super glad we get this, always wanted to play with PSO :)

File: src/orquestra/opt/optimizers/scipy_optimizer.py [hunk 8]
```diff
@@ -53,7 +53,7 @@ def __init__(
     def _minimize(
         self,
         cost_function: Union[CallableWithGradient, Callable],
-        initial_params: np.ndarray = None,
+        initial_params: np.ndarray,
```
[the remark this answers is not in the document]
Comment (VolodyaCO) [reply src/orquestra/opt/optimizers/scipy_optimizer.py:56]: Talking to @AthenaCaesura, we realised that it should be deleted. Initial parameters must be passed.

File: src/orquestra/opt/optimizers/pso/continuous_pso_optimizer.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (VolodyaCO) [reply src/orquestra/opt/optimizers/pso/continuous_pso_optimizer.py:40]: I have improved the output to be the union of (float, float) or (np.ndarray, np.ndarray). Basically, one uses those bounds to perform arithmetic operations, so broadcasting takes care of the case where the output is (float, float).
Since the function is used for velocity bounds as well (and we explicitly use those velocity bounds for some operations elsewhere), we should keep this function and not define `extract_scale_and_shift` as a complete block for treating the `Bounds` object.

File: src/orquestra/opt/optimizers/pso/continuous_pso_optimizer.py (hunk 2 above)
Comment (VolodyaCO) [reply src/orquestra/opt/optimizers/pso/continuous_pso_optimizer.py:154]: Solved

File: src/orquestra/opt/optimizers/pso/continuous_pso_optimizer.py (hunk 3 above)
Comment (VolodyaCO) [reply src/orquestra/opt/optimizers/pso/continuous_pso_optimizer.py:72]: Does sphinx render the type automatically from the signature?

File: src/orquestra/opt/optimizers/pso/continuous_pso_optimizer.py [hunk 9]
```diff
@@ -0,0 +1,333 @@
...
+        seed: Optional[int] = None,
+        recorder: RecorderFactory = _recorder,
+    ):
+        """
+        Constructor of the Particle Swarm Optimizer for continuous variables.
+        When you call `.minimize`, the initial positions of the particles can be
+        set using the method `get_initial_params`, which by default samples
+        uniformly from the parameter space.
+
+        Parameters
```
[the remark this answers is not in the document]
Comment (VolodyaCO) [reply src/orquestra/opt/optimizers/pso/continuous_pso_optimizer.py:70]: Done

File: src/orquestra/opt/optimizers/pso/continuous_pso_optimizer.py (hunk 4 above)
Comment (VolodyaCO) [reply src/orquestra/opt/optimizers/pso/continuous_pso_optimizer.py:183]: Done

File: src/orquestra/opt/optimizers/pso/continuous_pso_optimizer.py (hunk 5 above)
Comment (VolodyaCO) [reply src/orquestra/opt/optimizers/pso/continuous_pso_optimizer.py:186]: Not really because we need to store the velocity bounds. They are used elsewhere to clip the velocities.

File: src/orquestra/opt/optimizers/pso/continuous_pso_optimizer.py (hunk 6 above)
Comment (VolodyaCO) [reply src/orquestra/opt/optimizers/pso/continuous_pso_optimizer.py:164]: All functions and methods are now private

File: src/orquestra/opt/optimizers/pso/continuous_pso_optimizer.py (hunk 7 above)
Comment (VolodyaCO) [reply src/orquestra/opt/optimizers/pso/continuous_pso_optimizer.py:291]: Updated

File: src/orquestra/opt/optimizers/pso/continuous_pso_optimizer.py (hunk 3 above)
Comment (mstechly) [reply src/orquestra/opt/optimizers/pso/continuous_pso_optimizer.py]: Well, I'm not entirely sure.
If not, then we need to refactor our whole codebase anyway 🙃 
I just checked in our docs and I'm not sure what I'm seeing there, so it requires some extra investigation.

**Review by mstechly — COMMENTED** [review]
Code looks good, but tests and style don't pass :(

**Comment by VolodyaCO** [comment]
@mstechly as far as I understand, the style check is failing due to some stuff in the recorder? But I think I didn't touch that part of the code. As per the tests, it looks like PSO is not able to minimise some of the test functions, at least in the configuration of PSO hyperparameters that I'm testing!

**Comment by dexter2206** [comment]
> @mstechly as far as I understand, the style check is failing due to some stuff in the recorder? But I think I didn't touch that part of the code.

The error stems from this commit: [b3f4e74](https://github.com/zapatacomputing/orquestra-opt/pull/26/commits/b3f4e744ceaf059a2a499ee8361c26a434adcc44)

Long story short, what happened is that in some contexts it made `CallableStoringArtifacts` indistinguishable from the one that does not store artifacts. This is not bad per se. However, now some signatures of overloads for `recorder` overlap, but the return types don't have the same relationship as the arguments. 

To illustrate on a shorter example, consider a function with the following overloads (all capital letters are some types).

```python
@overload
def foo(x: S) -> T:
    pass

@overload
def foo(x: U) -> V:
    pass
```

If neither one of `S` and `U` are subtype of another, those overloads are perfectly independent. However, suppose `S` is a subtype of `U` and let `y` be of type `S`. What type is `foo(y)`? On the one hand, it has to be of type `T`, because of the first overload. However, it also has to be of type `V` because `y` being of type `S` also means that it is also of type `U`. What follows is that for those annotations to work we need to have "`T` is a subtype of `V`",

Back to our case. Consider the following overloads:
```python
@overload
def recorder(
    function: CallableWithGradientStoringArtifacts,
    save_condition: SaveCondition = always,
) -> ArtifactRecorderWithGradient:
    """The recorder function: variant for artifact-storing callables with gradient."""

@overload
def recorder(
    function: CallableWithGradient, save_condition: SaveCondition = always
) -> SimpleRecorderWithGradient:
    """The recorder function: variant for callables with gradient that don't store
    artifacts."""
```

Previously they were independent because `CallableWithGradientStoringArtifacts` was not a subtype of `CallableWithgradient` (and vice versa). However, after  [b3f4e74](https://github.com/zapatacomputing/orquestra-opt/pull/26/commits/b3f4e744ceaf059a2a499ee8361c26a434adcc44) `CallableWithGradientStoringArtifacts` is a subtype of `CallableWithGradient`. Hence, we have to carry this relationship over to `ArtifactRecorderWithGradient` and `SimpleRecorderWithGradient`, but currently, it isn't the case and it is precisely what mypy complains about:
```text
src/orquestra/opt/history/recorder.py:229: error: Overloaded function signatures 1 and 3 overlap with incompatible return types  [misc]

```

**Comment by mstechly** [comment]
@dexter2206 we don't have time right now for creating a new interface, could you do whatever fixes the problem fastest (option B if it's not much effort, if it is, please call our old friend). 
Feel free to add option A to the backlog if you think it's worthwhile :) 
Could you pair with @VolodyaCO on that, as he's using his own fork, it will probably be easiest to do this synchronously?

**Comment by VolodyaCO** [comment]
@dexter2206 @mstechly this is the problem arising by reverting the commit mentioned above:
```
src/orquestra/opt/optimizers/pso/continuous_pso_optimizer.py:273: error: Missing positional argument "store_artifact" in call to "__call__" of "CallableStoringArtifacts"  [call-arg]```

**Comment by dexter2206** [comment]
> @dexter2206 @mstechly this is the problem arising by reverting the commit mentioned above:
> 
> ```
> src/orquestra/opt/optimizers/pso/continuous_pso_optimizer.py:273: error: Missing positional argument "store_artifact" in call to "__call__" of "CallableStoringArtifacts"  [call-arg]```
> ```

I see what the problem is, and the typing hierarchy in optimizers is really messed up. Other optimizers use something like this instead of `CostFunction`:
```python
Union[CallableWithGradient, Callable]
```

So, revert the offending commit, change the type hint of `PSOOptimizer._minimize` to not use `CostFunction` but the union above, and everything will run smoothly.

**Comment by VolodyaCO** [comment]
Ok, I'll push this solution @dexter2206. However, `_minimize` should accept a `CostFunction` instance for the `cost_function` argument, as per the `Optimizer._minimize` abstract method signature.

File: src/orquestra/opt/optimizers/pso/continuous_pso_optimizer.py (hunk 3 above)
Comment (mstechly) [reply src/orquestra/opt/optimizers/pso/continuous_pso_optimizer.py]: Ok, could you change the style of the docs to be consistent with how it's being formatted in other places?

Comment (VolodyaCO) [reply src/orquestra/opt/optimizers/pso/continuous_pso_optimizer.py]: Actually, I had already eliminated the types in the docstrings as well! So this has been dealt with 😄

Comment (mstechly) [reply src/orquestra/opt/optimizers/pso/continuous_pso_optimizer.py]: Thank you, merging it then!

**Review by mstechly — APPROVED** [review]
